### PR TITLE
Remove auth for procedures

### LINKED ISSUE #233: Cannot invoke kind-level procedures when logged in with a real token
Getting the following error:
```
papiea-engine    | CasbinAuthorizer checkPermission error TypeError: Cannot read property 'kind' of undefined
papiea-engine    |     at evaluateMember (/code/papiea-engine/node_modules/expression-eval/index.js:48:27)
papiea-engine    |     at evaluate (/code/papiea-engine/node_modules/expression-eval/index.js:94:14)
papiea-engine    |     at evaluate (/code/papiea-engine/node_modules/expression-eval/index.js:60:71)
papiea-engine    |     at evaluate (/code/papiea-engine/node_modules/expression-eval/index.js:87:16)
papiea-engine    |     at evaluate (/code/papiea-engine/node_modules/expression-eval/index.js:89:50)
papiea-engine    |     at evaluate (/code/papiea-engine/node_modules/expression-eval/index.js:89:16)
papiea-engine    |     at Enforcer.enforce (/code/papiea-engine/node_modules/casbin/lib/coreEnforcer.js:279:32)
papiea-engine    |     at CasbinAuthorizer.checkPermission (/code/papiea-engine/src/auth/casbin.ts:32:32)
papiea-engine    |     at PerProviderAuthorizer.checkPermission (/code/papiea-engine/src/auth/authz.ts:156:27)
papiea-engine    |     at processTicksAndRejections (internal/process/task_queues.js:85:5)
```

Trying to look deeper into it, in here https://github.com/nutanix/papiea-js/blob/13ecded8e2a10337fcfafdf2532e1ee94f257c8c/papiea-engine/src/auth/casbin.ts#L26-L36

Printing `user`, `object` and `action` gives the following:
1. `user`:
```
papiea-engine    | SHLOMI:::: user
{  
   "tenant":"1c7fb69f-ed02-4e47-9881-79e8f0bf90ca",
   "tenant-email":"shlomi.vaknin@nutanix.com",
   "tenant-fname":"Shlomi",
   "tenant-lname":"Vaknin",
   "tenant-role":"account-admin",
   "owner":"shlomi.vaknin@nutanix.com",
   "authorization":"Bearer eyJhbGciOiJS...DrB-w",
   "provider_prefix":"xi_clusters",
   "provider_version":"0.0.1",
   "iat":1563399927,
   "exp":1564004727
}
```
2. `object`:
```
papiea-engine    | SHLOMI:::: object
{  
   "kind":{  
      "name":"imaged_cluster",
      "name_plural":"imaged_clusters",
      "kind_structure":{  
         "imaged_cluster":{  
            "type":"object",
            "title":"imaged_cluster",
            "description":"Stores an imaged cluster",
            "x-papiea-entity":"spec-only",
            "properties":{  
               "name":{  
                  "type":"string"
               },
               "number_of_hosts":{  
                  "type":"number"
               },
               "cvm_ip":{  
                  "x-papiea":"status-only",
                  "type":"string"
               },
               "host_reference_list":{  
                  "x-papiea":"status-only",
                  "type":"array",
                  "items":{  
                     "type":"object",
                     "required":[  
                        "uuid",
                        "kind"
                     ],
                     "properties":{  
                        "uuid":{  
                           "type":"string"
                        },
                        "name":{  
                           "type":"string"
                        },
                        "kind":{  
                           "type":"string"
                        }
                     }
                  }
               },
               "geolocation_reference":{  
                  "x-papiea":"status-only",
                  "type":"object",
                  "required":[  
                     "uuid",
                     "kind"
                  ],
                  "properties":{  
                     "uuid":{  
                        "type":"string"
                     },
                     "name":{  
                        "type":"string"
                     },
                     "kind":{  
                        "type":"string"
                     }
                  }
               },
               "cluster_status":{  
                  "x-papiea":"status-only",
                  "type":"string"
               }
            }
         }
      },
      "intentful_signatures":{  

      },
      "dependency_tree":{  

      },
      "kind_procedures":{  
         "create_aws_cluster":{  
            "name":"create_aws_cluster",
            "argument":{  
               "Create_AWS_Cluster_Input":{  
                  "type":"object",
                  "title":"Create AWS Cluster Input",
                  "description":"All info required to create a xi-clusters cluster on AWS bare metal hosts",
                  "properties":{  
                     "name":{  
                        "type":"string",
                        "title":"Name of the cluster",
                        "example":"New cluster"
                     },
                     "geolocation_reference":{  
                        "type":"object",
                        "title":"Geolocation reference",
                        "description":"Geolocation where xi-clusters should run in",
                        "required":[  
                           "uuid",
                           "kind"
                        ],
                        "properties":{  
                           "uuid":{  
                              "type":"string",
                              "example":"3fa85f64-5717-4562-b3fc-2c963f66afa6"
                           },
                           "kind":{  
                              "type":"string",
                              "example":"geolocation"
                           },
                           "name":{  
                              "type":"string",
                              "example":"entity_name"
                           }
                        }
                     },
                     "host_type_reference":{  
                        "type":"object",
                        "title":"Host type reference",
                        "description":"AWS instance type on which xi-clusters should run",
                        "required":[  
                           "uuid",
                           "kind"
                        ],
                        "properties":{  
                           "uuid":{  
                              "type":"string",
                              "example":"0ed241b6-d0a4-46fe-ab12-6f0b3ee0da7c"
                           },
                           "kind":{  
                              "type":"string",
                              "example":"host_type"
                           },
                           "name":{  
                              "type":"string",
                              "example":"entity_name"
                           }
                        }
                     },
                     "number_of_hosts":{  
                        "type":"number",
                        "title":"The requested number of hosts",
                        "example":3
                     },
                     "aos_version":{  
                        "type":"string",
                        "title":"AOS Version",
                        "example":"5.10.3.2"
                     },
                     "network":{  
                        "type":"object",
                        "title":"Cluster network configuration",
                        "properties":{  
                           "source_cidr":{  
                              "type":"string",
                              "title":"Allowed IPs cidr",
                              "description":"Only IPs belonging to this cidr will be allowed to connect to these hosts. This is essentially AWS security groups for the hosts",
                              "example":"0.0.0.0/0"
                           },
                           "use_default_subnet":{  
                              "type":"boolean",
                              "title":"Use the default VPC's default subnet",
                              "description":"Start the cluster in the region's default VPC default subnet, and assign a public ip to the cluster nodes",
                              "example":true
                           },
                           "autogenerate_uvms_private_subnet":{  
                              "type":"boolean",
                              "title":"Generate a private subnet for the UVMs",
                              "description":"If true, a new private subnet will be generated for the UVMs of this cluster. The private subnet will be routed through a NAT on the instance's public subnet to have Internet access",
                              "example":true
                           }
                        }
                     }
                  }
               }
            },
            "result":{  
               "Create_AWS_Cluster_Output":{  
                  "type":"object",
                  "title":"Create AWS Cluster Output",
                  "properties":{  
                     "cloud_formation_url":{  
                        "type":"string",
                        "title":"Cloudformation URL",
                        "description":"The URL to follow in order to start cluster creation in the user AES account"
                     },
                     "cluster_reference":{  
                        "type":"object",
                        "title":"Created Cluster reference",
                        "description":"The reference to the created imaged_cluster",
                        "required":[  
                           "uuid",
                           "kind"
                        ],
                        "properties":{  
                           "uuid":{  
                              "type":"string",
                              "example":"entity_uuid"
                           },
                           "kind":{  
                              "type":"string",
                              "example":"kind_name"
                           },
                           "name":{  
                              "type":"string",
                              "example":"entity_name"
                           }
                        }
                     }
                  }
               }
            },
            "execution_strategy":0,
            "procedure_callback":"http://docker.for.mac.localhost:9999/imaged_cluster/create_aws_cluster"
         }
      },
      "entity_procedures":{  
         "rename":{  
            "name":"rename",
            "argument":{  

            },
            "result":{  

            },
            "execution_strategy":0,
            "procedure_callback":"http://docker.for.mac.localhost:9999/imaged_cluster/rename"
         }
      }
   }
}
```
3. `action`:
```
papiea-engine    | SHLOMI:::: action "call_create_aws_cluster"
```
And true enough, the object does not have a `metadata.kind` in it (as described in the model).

Is this the intended object that is checked?

p.s. 
If I use an admin s2s key (either papiea or provider), this seems to work fine, probably this entire step is not executed.

### PR DESCRIPTION
**Pull request by adolgarev**
Resolves #233

### CODE REVIEW

File: papiea-engine/__tests__/auth.yaml [hunk 1]
```diff
@@ -42,5 +42,4 @@ oauth:
 
       # role will be populated by what `role` got resolved into in its section above
       tenant-role: ^role
-      owner: ^id_token.sub
-      authorization: $bearer(^token.access_token)
```
Comment (shlomi-nx) [inline papiea-engine/__tests__/auth.yaml]: Why did you remove authorization from the headers?

File: papiea-engine/__tests__/entity_api_auth.test.ts [hunk 2]
```diff
@@ -431,10 +446,19 @@ describe("Entity API auth tests", () => {
                 });
                 req.on('end', function () {
                     const post = JSON.parse(body);
-                    res.statusCode = 200;
-                    res.setHeader('Content-Type', 'text/plain');
-                    res.end(JSON.stringify(post.spec));
-                    server.close();
+                    const authorization = req.headers['authorization'];
```
Comment (shlomi-nx) [inline papiea-engine/__tests__/entity_api_auth.test.ts]: And how are you using it here?

**Comment by shlomi-nx** [comment]
@adolgarev @Emulebest Bearer should always be the token received from the oauth2 server, not the papiea token. Papiea may add more fields through the same userInfo mechanism.

File: papiea-engine/src/auth/authn.ts [hunk 3]
```diff
@@ -10,6 +15,108 @@ export class UnauthorizedError extends Error {
...
+            console.log(JSON.parse(atob(token)));
+            const provider: Provider = await this.providerDb.get_provider(this.provider_prefix, this.provider_version);
+            const userInfo = getUserInfoFromToken(JSON.parse(atob(token)), provider);
+            userInfo.provider_prefix = this.provider_prefix;
+            userInfo.provider_version = this.provider_version;
+            console.log(userInfo.authorization);
+            delete userInfo.is_admin;
+            return userInfo;
+        } catch (e) {
+            console.error(e);
```
Comment (shlomi-nx) [inline papiea-engine/src/auth/authn.ts]: Please add some text for debugging context

File: papiea-engine/src/auth/authn.ts [hunk 4]
```diff
@@ -10,6 +15,108 @@ export class UnauthorizedError extends Error {
...
+    }
+
+    async getUserAuthInfo(token: string): Promise<UserAuthInfo | null> {
+        try {
+            if (!this.provider_prefix || !this.provider_version) {
+                return null;
+            }
+            console.log(JSON.parse(atob(token)));
+            const provider: Provider = await this.providerDb.get_provider(this.provider_prefix, this.provider_version);
+            const userInfo = getUserInfoFromToken(JSON.parse(atob(token)), provider);
```
Comment (shlomi-nx) [inline papiea-engine/src/auth/authn.ts]: Why did you add the `atob` and `JSON.parse` here and in other places?

**Comment by shlomi-nx** [comment]
@Emulebest two points:
1. Lets rename all `extension` of s2s_keys to `user_info`
2. s2s keys should have `provider_prefix` inside of it, near `name?` and `key` - not in the extension:
    https://github.com/nutanix/papiea-js/blob/467e78ec0d8534409c736e044123d77ffa9d4142/papiea-core/src/core.ts#L197-L207
    (`extension: UserInfo` should turn into `user_info:UserInfo`)

**Comment by shlomi-nx** [comment]
You can see how this is being used here https://github.com/nutanix/nunet-ui/pull/304:

https://github.com/nutanix/nunet-ui/blob/f6cb860e0bc05a89053e190e841558527a4034ec/xiclusters-provider/src/kinds/imaged_cluster.ts#L74-L132

File: papiea-engine/src/auth/authn.ts (hunk 4 above)
Comment (Emulebest) [reply papiea-engine/src/auth/authn.ts]: We get a JS object as a representation of IDP token. To send it over internet as a part of headers we need string, so -> JSON.stringify, after that we need to convert it to base64 for convenient use -> btoa. To read the token we need JSON.parse(atob(token))

Comment (shlomi-nx) [reply papiea-engine/src/auth/authn.ts]: Oh right, thanks

**Comment by shlomi-nx** [comment]
> @Emulebest two points:
> 
> 1. Lets rename all `extension` of s2s_keys to `user_info`
> 2. s2s keys should have `provider_prefix` inside of it, near `name?` and `key` - not in the extension:
>    https://github.com/nutanix/papiea-js/blob/467e78ec0d8534409c736e044123d77ffa9d4142/papiea-core/src/core.ts#L197-L207
>    
>    (`extension: UserInfo` should turn into `user_info:UserInfo`)
> 3. Need to improve returning errors from a procedure. Need to signal what kind of error this is,(unauthorized, bad request, etc.) Need to have either automatic propagation of errors to the client, or have the client receive an opaque error but have the provider logs describe the context of the errors (See what I did in the procedure with `e.response.data.errors`)

If this is done, including the three points above, please merge and close.
